XML Catalog with p:validate-with-xml-schema not working as expected

Perhaps (likely) this is a XML Catalog usage issue, but XMLCalabash is my best window into it. 

I have a schema, which references a number of other schemas. I am able invoke `p:validate-with-xml-schema` and it will pass in the correct schema from the XML Catalog. However, the other schemas referenced using `xsd:import` are not being located correctly. 

Saxon-EE, called directly performs validation as expected.

E.g.

`"$JAVA_HOME/bin/java" -Dxml.catalog.defaultLoggerLogLevel=debug -Dxml.catalog.logging=request:info,response:info -Dxml.catalog.catalogLoaderClass=org.xmlresolver.loaders.ValidatingXmlLoader -cp "$CP" com.saxonica.Validate -xsd:./catalog_documents/test_import.xsd -xsdversion:1.0 -catalog:./config/repro_catalog.xml ./resources/example_repro.xml`

From what I can tell, this is because while the catalog will be used to locate the schema input from the XProc step, the catalog is not being used by the `p:validate-with-xml-schema` step itself, and thus tries to deference internet URIs that will never resolve.

`"$JAVA_HOME/bin/java" -Dxml.catalog.defaultLoggerLogLevel=debug -Dxml.catalog.logging=request:info,response:info -Dxml.catalog.files="./config/repro_catalog.xml;" -Dxml.catalog.prefer=true -Dxml.catalog.uriForSystem=true -cp "$CP" com.xmlcalabash.app.Main  ./tests/import_catalog_repro.xpl --verbosity:debug --config:./config/calabash.xml`

Because there are intractably many files involved, I will send you a reproduction environment separately.

STDOUT:

```
╰─$ ./xmlcalabash-3.0.0-alpha20/repro-catalog-xmlcalabash.sh ./tests/import_catalog_repro.xpl --verbosity:debug --config:./config/calabash.xml
{repeating config sections removed...}
config: Logging: request:info,response:info
config: Prefer public: false
config: Prefer property file: false
config: Allow catalog PI: true
config: Always resolve: true
config: Parse RDDL: true
config: URI for system: true
config: Merge http/https: true
config: Mask jar URIs: true
config: Catalog loader: org.xmlresolver.loaders.XmlLoader
config: Classpath catalogs: true
config: Archived catalogs: true
config: Throw URI exceptions: false
config: Class loader: jdk.internal.loader.ClassLoaders$AppClassLoader@639fee48
config: Logger class: org.xmlresolver.logging.DefaultLogger
config: Access external entity: all
config: Access external document: all
config: SAXParserFactory class: null
config: XMLReader supplier: org.xmlresolver.ResolverFeature$$Lambda/0x00007fe974087b58@3d299e3
config: Fix Windows system identifiers: false
config: Default logger log level: debug
config: Catalog: jar:file:/c/Users/John/Documents/GitHub/cdf-test-method-nv/xmlcalabash-3.0.0-alpha20/lib/xmlresolver-6.0.12-data.jar!/org/xmlresolver/catalog.xml
config: Throw URI exceptions: true
Debug: Running p:declare-step/!declare-step (declare-step)
Debug: Running (compound step head declare-step_head)
Debug: Running cx:document/!document (document)
request: lookupUri: http://example.com/import (baseURI: null)
Debug: Running cx:inline/!inline (inline)
Debug: Running p:validate-with-xml-schema/val (val)
Debug: Caching input schema file:/c/Users/John/Documents/GitHub/cdf-test-method-nv/catalog_documents/test_import.xsd for http://example.com/import
request: lookupUri: http://example.com/base (baseURI: file:/c/Users/John/Documents/GitHub/cdf-test-method-nv/catalog_documents/test_import.xsd)
config: Loaded catalog: jar:file:/c/Users/John/Documents/GitHub/cdf-test-method-nv/xmlcalabash-3.0.0-alpha20/lib/xmlresolver-6.0.12-data.jar!/org/xmlresolver/catalog.xml
request: lookupEntity: http://example.com/base (baseURI: file:/c/Users/John/Documents/GitHub/cdf-test-method-nv/catalog_documents/test_import.xsd, publicId: null)
warning: Failed to GET: http://example.com/base: http://example.com/base
Debug: Imported schema document http://example.com/base cannot be located: java.io.FileNotFoundException (http://example.com/base)
Debug: Unknown type Q{http://example.com/base}PersonType
Debug: The type {PersonType} is referenced, but has not been declared
Fatal err:XC0152 at file:/c/Users/John/Documents/GitHub/cdf-test-method-nv/./tests/import_catalog_repro.xpl:3:68: Invalid XML Schema: “file:/c/Users/John/Documents/GitHub/cdf-test-method-nv/catalog_documents/test_import.xsd”.

com.xmlcalabash.exceptions.XProcException: err:XC0152: file:/c/Users/John/Documents/GitHub/cdf-test-method-nv/catalog_documents/test_import.xsd
        at com.xmlcalabash.runtime.XProcStepConfigurationImpl.exception(XProcStepConfigurationImpl.kt:244)
        at com.xmlcalabash.steps.validation.ValidateWithXmlSchema.validateWithSaxon(ValidateWithXmlSchema.kt:149)
        at com.xmlcalabash.steps.validation.ValidateWithXmlSchema.run(ValidateWithXmlSchema.kt:52)
        at com.xmlcalabash.runtime.steps.AtomicStep.runImplementation$xmlcalabash(AtomicStep.kt:155)
        at com.xmlcalabash.runtime.steps.AtomicStep.run(AtomicStep.kt:146)
        at com.xmlcalabash.runtime.steps.AbstractStep.runStep(AbstractStep.kt:149)
        at com.xmlcalabash.runtime.steps.CompoundStep.runStepsExhaustively(CompoundStep.kt:166)
        at com.xmlcalabash.runtime.steps.CompoundStep.runSubpipeline(CompoundStep.kt:112)
        at com.xmlcalabash.runtime.steps.GroupStep.run(GroupStep.kt:15)
        at com.xmlcalabash.runtime.steps.AbstractStep.runStep(AbstractStep.kt:149)
        at com.xmlcalabash.runtime.steps.PipelineStep.runStep(PipelineStep.kt:36)
        at com.xmlcalabash.runtime.XProcPipeline.run(XProcPipeline.kt:140)
        at com.xmlcalabash.app.XmlCalabashCli.run(XmlCalabashCli.kt:304)
        at com.xmlcalabash.app.XmlCalabashCli.access$run(XmlCalabashCli.kt:64)
        at com.xmlcalabash.app.XmlCalabashCli$Companion.run(XmlCalabashCli.kt:68)
        at com.xmlcalabash.app.Main$Companion.main(Main.kt:12)
        at com.xmlcalabash.app.Main.main(Main.kt)
Caused by: net.sf.saxon.s9api.SaxonApiException: Errors were found in the schema
        at com.saxonica.ee.s9api.SchemaManagerImpl.load(SchemaManagerImpl.java:164)
        at com.xmlcalabash.steps.validation.ValidateWithXmlSchema.validateWithSaxon(ValidateWithXmlSchema.kt:146)
        ... 15 more
Caused by: net.sf.saxon.type.SchemaException: Errors were found in the schema
        at com.saxonica.ee.schema.PreparedSchema.fixupComponentReferences(PreparedSchema.java:258)
        at com.saxonica.ee.schema.PreparedSchema.validate(PreparedSchema.java:1005)
        at com.saxonica.ee.schema.SchemaCompiler.compile(SchemaCompiler.java:496)
        at com.saxonica.ee.schema.SchemaCompiler.validate(SchemaCompiler.java:530)
        at com.saxonica.config.EnterpriseConfiguration.addSchemaSource(EnterpriseConfiguration.java:650)
        at com.saxonica.config.EnterpriseConfiguration.addSchemaSource(EnterpriseConfiguration.java:612)
        at com.saxonica.ee.s9api.SchemaManagerImpl.load(SchemaManagerImpl.java:162)
        ... 16 more
net.sf.saxon.s9api.SaxonApiException: Errors were found in the schema
        at com.saxonica.ee.s9api.SchemaManagerImpl.load(SchemaManagerImpl.java:164)
        at com.xmlcalabash.steps.validation.ValidateWithXmlSchema.validateWithSaxon(ValidateWithXmlSchema.kt:146)
        at com.xmlcalabash.steps.validation.ValidateWithXmlSchema.run(ValidateWithXmlSchema.kt:52)
        at com.xmlcalabash.runtime.steps.AtomicStep.runImplementation$xmlcalabash(AtomicStep.kt:155)
        at com.xmlcalabash.runtime.steps.AtomicStep.run(AtomicStep.kt:146)
        at com.xmlcalabash.runtime.steps.AbstractStep.runStep(AbstractStep.kt:149)
        at com.xmlcalabash.runtime.steps.CompoundStep.runStepsExhaustively(CompoundStep.kt:166)
        at com.xmlcalabash.runtime.steps.CompoundStep.runSubpipeline(CompoundStep.kt:112)
        at com.xmlcalabash.runtime.steps.GroupStep.run(GroupStep.kt:15)
        at com.xmlcalabash.runtime.steps.AbstractStep.runStep(AbstractStep.kt:149)
        at com.xmlcalabash.runtime.steps.PipelineStep.runStep(PipelineStep.kt:36)
        at com.xmlcalabash.runtime.XProcPipeline.run(XProcPipeline.kt:140)
        at com.xmlcalabash.app.XmlCalabashCli.run(XmlCalabashCli.kt:304)
        at com.xmlcalabash.app.XmlCalabashCli.access$run(XmlCalabashCli.kt:64)
        at com.xmlcalabash.app.XmlCalabashCli$Companion.run(XmlCalabashCli.kt:68)
        at com.xmlcalabash.app.Main$Companion.main(Main.kt:12)
        at com.xmlcalabash.app.Main.main(Main.kt)
Caused by: net.sf.saxon.type.SchemaException: Errors were found in the schema
        at com.saxonica.ee.schema.PreparedSchema.fixupComponentReferences(PreparedSchema.java:258)
        at com.saxonica.ee.schema.PreparedSchema.validate(PreparedSchema.java:1005)
        at com.saxonica.ee.schema.SchemaCompiler.compile(SchemaCompiler.java:496)
        at com.saxonica.ee.schema.SchemaCompiler.validate(SchemaCompiler.java:530)
        at com.saxonica.config.EnterpriseConfiguration.addSchemaSource(EnterpriseConfiguration.java:650)
        at com.saxonica.config.EnterpriseConfiguration.addSchemaSource(EnterpriseConfiguration.java:612)
        at com.saxonica.ee.s9api.SchemaManagerImpl.load(SchemaManagerImpl.java:162)
        ... 16 more
```

XML Catalog:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE catalog PUBLIC "-//OASIS//DTD XML Catalogs V1.1//EN" "http://www.oasis-open.org/committees/entity/release/1.1/catalog.dtd">
<catalog xmlns="urn:oasis:names:tc:entity:xmlns:xml:catalog">
    <uri name="http://example.com/base" uri="../catalog_documents/test_base.xsd" />
    <uri name="http://example.com/import" uri="../catalog_documents/test_import.xsd" />
</catalog>
```

XProc:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<?xml-model href="../resources/xproc30.rng" type="application/xml"?>
<p:declare-step xmlns:p="http://www.w3.org/ns/xproc" version="3.0">
    <p:output port="result" content-types="any" pipe="report@val" />
    <p:validate-with-xml-schema name="val">
        <p:with-input port="schema" href="http://example.com/import" />
        <p:with-input port="source" href="../resources/example_repro.xml" />
    </p:validate-with-xml-schema>
</p:declare-step>
```

--- a/xmlcalabash/src/main/kotlin/com/xmlcalabash/io/DocumentManager.kt
+++ b/xmlcalabash/src/main/kotlin/com/xmlcalabash/io/DocumentManager.kt
@@ -6,14 +6,18 @@ import com.xmlcalabash.documents.XProcDocument
 import com.xmlcalabash.exceptions.XProcError
 import com.xmlcalabash.runtime.XProcStepConfiguration
 import com.xmlcalabash.spi.DocumentResolver
+import net.sf.saxon.Configuration
 import net.sf.saxon.lib.ModuleURIResolver
 import net.sf.saxon.lib.ResourceResolver
+import net.sf.saxon.lib.SchemaURIResolver
 import net.sf.saxon.s9api.QName
 import net.sf.saxon.s9api.XdmValue
 import org.apache.logging.log4j.kotlin.logger
 import org.xml.sax.EntityResolver
 import org.xml.sax.InputSource
 import org.xml.sax.ext.EntityResolver2
+import org.xmlresolver.Resolver
+import org.xmlresolver.ResolverConstants
 import org.xmlresolver.ResourceRequestImpl
 import org.xmlresolver.XMLResolver
 import org.xmlresolver.XMLResolverConfiguration
@@ -203,7 +207,7 @@ class DocumentManager(): EntityResolver, EntityResolver2, ResourceResolver, Modu
         val request = ResourceRequestImpl(resolverConfiguration, saxonRequest.nature, saxonRequest.purpose)
         request.uri = saxonRequest.uri
         request.baseURI = saxonRequest.baseUri
-        request.encoding = saxonRequest.entityName
+        request.entityName = saxonRequest.entityName
         request.publicId = saxonRequest.publicId
 
         val resp = resolver.resolve(request)
@@ -230,6 +234,16 @@ class DocumentManager(): EntityResolver, EntityResolver2, ResourceResolver, Modu
 
         val inputSource = inputFromCache(href)
         if (inputSource == null) {
+            val request = ResourceRequestImpl(resolverConfiguration, ResolverConstants.SCHEMA_NATURE, ResolverConstants.VALIDATION_PURPOSE)
+            request.uri = moduleURI
+            request.baseURI = baseURI
+
+            val resp = resolver.resolve(request)
+            if (resp.inputStream != null) {
+                val source = StreamSource(resp.inputStream, href.toString())
+                return arrayOf(source)
+            }
+
             return null
         }
 

--- a/xmlcalabash/src/main/kotlin/com/xmlcalabash/steps/validation/ValidateWithXmlSchema.kt
+++ b/xmlcalabash/src/main/kotlin/com/xmlcalabash/steps/validation/ValidateWithXmlSchema.kt
@@ -12,13 +12,17 @@ import com.xmlcalabash.runtime.parameters.RuntimeStepParameters
 import com.xmlcalabash.steps.AbstractAtomicStep
 import com.xmlcalabash.util.S9Api
 import com.xmlcalabash.util.SaxonErrorReporter
+import net.sf.saxon.Configuration
 import net.sf.saxon.Controller
+import net.sf.saxon.lib.SchemaURIResolver
 import net.sf.saxon.om.NamespaceUri
 import net.sf.saxon.s9api.*
 import net.sf.saxon.serialize.SerializationProperties
 import net.sf.saxon.type.ValidationException
 import java.net.URI
+import javax.xml.transform.Source
 import javax.xml.transform.sax.SAXSource
+import javax.xml.transform.stream.StreamSource
 
 open class ValidateWithXmlSchema(): AbstractAtomicStep() {
     companion object {
@@ -48,7 +52,10 @@ open class ValidateWithXmlSchema(): AbstractAtomicStep() {
         if (manager == null) {
             throw RuntimeException("Schema manager not found, XSD validation requires Saxon EE")
         }
+
         manager.errorReporter = errorReporter
+        manager.schemaURIResolver = XsdResolver()
+
         validateWithSaxon(manager)
     }
 
@@ -227,4 +234,14 @@ open class ValidateWithXmlSchema(): AbstractAtomicStep() {
     }
 
     override fun toString(): String = "p:validate-with-xml-schema"
+
+    inner class XsdResolver(): SchemaURIResolver {
+        override fun setConfiguration(config: Configuration?) {
+            // I don't care???
+        }
+
+        override fun resolve(moduleURI: String?, baseURI: String?, locations: Array<out String>?): Array<StreamSource>? {
+            return stepConfig.environment.documentManager.resolve(moduleURI, baseURI, locations)
+        }
+    }
 }

